lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -348,21 +348,52 @@ static void collectLoads(SILInstruction *I, SmallVectorImpl<LoadInst *> &Loads)
 static void replaceLoad(LoadInst *LI, SILValue val, AllocStackInst *ASI) {
   ProjectionPath projections(val->getType());
   SILValue op = LI->getOperand();
+  SILBuilderWithScope builder(LI);
+
   while (op != ASI) {
     assert(isa<UncheckedAddrCastInst>(op) || isa<StructElementAddrInst>(op) ||
            isa<TupleElementAddrInst>(op));
     auto *Inst = cast<SingleValueInstruction>(op);
     projections.push_back(Projection(Inst));
     op = Inst->getOperand(0);
   }
-  SILBuilder builder(LI);
+
+  SmallVector<SILValue, 4> borrowedVals;
   for (auto iter = projections.rbegin(); iter != projections.rend(); ++iter) {
     const Projection &projection = *iter;
+    assert(projection.getKind() == ProjectionKind::BitwiseCast ||
+           projection.getKind() == ProjectionKind::Struct ||
+           projection.getKind() == ProjectionKind::Tuple);
+
+    // struct_extract and tuple_extract expect guaranteed operand ownership
+    // non-trivial RunningVal is owned. Insert borrow operation to convert
+    if (projection.getKind() == ProjectionKind::Struct ||
+        projection.getKind() == ProjectionKind::Tuple) {
+      SILValue opVal = builder.emitBeginBorrowOperation(LI->getLoc(), val);
+      if (opVal != val) {
+        borrowedVals.push_back(opVal);
+        val = opVal;
+      }
+    }
     val = projection.createObjectProjection(builder, LI->getLoc(), val).get();
   }
+
   op = LI->getOperand();
-  LI->replaceAllUsesWith(val);
+  // Replace users of the loaded value with `val`
+  // If we have a load [copy], replace the users with copy_value of `val`
+  if (LI->getOwnershipQualifier() == LoadOwnershipQualifier::Copy) {
+    LI->replaceAllUsesWith(builder.createCopyValue(LI->getLoc(), val));
+  } else {
+    LI->replaceAllUsesWith(val);
```

Comment:
I think this is probably incorrect if we have a load [take] and a struct_extract/tuple_extract. In that case we will have a guaranteed value being passed off as an owned value. I think this is a missing test case. We probably need to insert a copy_value here, no?

Reply:
Looking..

Reply:
Or maybe we need to do something sort of like PMO where we eliminate copies first and then try to eliminate takes + the whole allocation later? Your thoughts?

Reply:
Is there any case where we would have a guaranteed value here? I don't think that's possible because the value we load must have been stored in the stack alloc at one point which means it must be an owned value.

Reply:
Right. `val` is owned here. borrow scope will end before the uses of loaded value.

Reply:
Actually gottesmm is right. Loaded val is owned, but the result of struct_extract will be guaranteed

Reply:
But, we won't optimize something if there is a consuming use of a `struct_element_addr`. A use of a `struct_element_addr` has to be a load/store into the original stack alloc (which could never happen because they wouldn't be the correct type), or it would have to be an unsupported instruction. In other words, I think we would bail if there was a consuming use of a loaded `struct_element_addr`. Does that logic make sense?

Reply:
@zoecarver It can happen. See `mem2reg_struct_addr_load_take` in `mem2reg_ossa_nontrivial.sil`

I disabled this pattern `(load [take] (struct_element_addr/tuple_element_addr %ASI))` for now. We'll need a copy_value like gottesmm said. But we also need destroy of the original `RunningVal`. I started implementing it, but it started getting too complex and I don't want to add that in this PR.

@gottesmm thanks for catching this!

Reply:
Ah, you're right. I was thinking that `isCaptured` would reject that function but, it doesn't (which is obvious now that I have the example in front of me). 

> I started implementing it, but it started getting too complex and I don't want to add that in this PR.

This is where the `ValueLifetimeAnalysis` might come in handy.